platform/machine/unprivqemu/cluster.go

```diff
@@ -0,0 +1,197 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unprivqemu
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pborman/uuid"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/conf"
+	"github.com/coreos/mantle/system/exec"
+	"github.com/coreos/mantle/util"
+)
+
+// Cluster is a local cluster of QEMU-based virtual machines.
+//
+// XXX: must be exported so that certain QEMU tests can access struct members
+// through type assertions.
+type Cluster struct {
+	*platform.BaseCluster
+	flight *flight
+
+	mu sync.Mutex
+}
+
+func (qc *Cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
+	return qc.NewMachineWithOptions(userdata, platform.MachineOptions{})
+}
+
+func (qc *Cluster) NewMachineWithOptions(userdata *conf.UserData, options platform.MachineOptions) (platform.Machine, error) {
+	id := uuid.New()
+
+	dir := filepath.Join(qc.RuntimeConf().OutputDir, id)
+	if err := os.Mkdir(dir, 0777); err != nil {
+		return nil, err
+	}
+
+	// hacky solution for cloud config ip substitution
+	// NOTE: escaping is not supported
+	qc.mu.Lock()
+
+	conf, err := qc.RenderUserData(userdata, map[string]string{})
+	if err != nil {
+		qc.mu.Unlock()
+		return nil, err
+	}
+	qc.mu.Unlock()
+
+	var confPath string
+	if conf.IsIgnition() {
+		confPath = filepath.Join(dir, "ignition.json")
+		if err := conf.WriteFile(confPath); err != nil {
+			return nil, err
+		}
+	} else if conf.IsEmpty() {
+	} else {
+		return nil, fmt.Errorf("unprivileged qemu only supports Ignition or empty configs")
+	}
+
+	journal, err := platform.NewJournal(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	qm := &machine{
+		qc:          qc,
+		id:          id,
+		journal:     journal,
+		consolePath: filepath.Join(dir, "console.txt"),
+	}
+
+	qmCmd, extraFiles, err := platform.CreateQEMUCommand(qc.flight.opts.Board, qm.id, qc.flight.opts.BIOSImage, qm.consolePath, confPath, qc.flight.diskImagePath, conf.IsIgnition(), options)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range extraFiles {
+		defer file.Close()
+	}
+
+	qc.mu.Lock()
+
+	qmCmd = append(qmCmd, "-netdev", "user,id=eth0,restrict=yes,hostfwd=tcp:127.0.0.1:0-:22", "-device", platform.Virtio(qc.flight.opts.Board, "net", "netdev=eth0"))
+
+	plog.Debugf("NewMachine: %q", qmCmd)
+
+	qm.qemu = exec.Command(qmCmd[0], qmCmd[1:]...)
+
+	qc.mu.Unlock()
+
+	cmd := qm.qemu.(*exec.ExecCmd)
+	cmd.Stderr = os.Stderr
+
+	cmd.ExtraFiles = append(cmd.ExtraFiles, extraFiles...)
+
+	if err = qm.qemu.Start(); err != nil {
+		return nil, err
+	}
+
+	// parse /proc/net/tcp to determine the port selected by QEMU
+	getAddress := func() (string, error) {
```

Comment:
is there a reason for this being a closure instead of a normal function?

Reply:
No particular reason, was originally just in-function code but ended up needing to be inside of a sub-function for `util.Retry` so I just wrapped it into a closure rather than creating a separate function and passing through `qm.qemu.Pid()`

Reply:
I'd rather have it as a function in that case, make it explicit we're not trying to capture variables or anything fancy.